### PR TITLE
Phase7 - 7.3 cacheable board read service layer

### PR DESCRIPTION
**Pull request by kanghowoo**
When getting a board, if the cache does not contain the data, it will query the database, and then cache the retrieved data using the `cacheBoard()` method.
This method is placed in `CacheableBoardReadServiceImpl`
Do you think this placement is appropriate?

### CODE REVIEW

File: src/main/java/com/mide/gangsaeng/board/BoardReadService.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.mide.gangsaeng.board;
+
+public interface BoardReadService {
```
Comment (TK-one) [inline src/main/java/com/mide/gangsaeng/board/BoardReadService.java:3]: Why did you create new `BoardReadService`?

Comment (kanghowoo) [reply src/main/java/com/mide/gangsaeng/board/BoardReadService.java:3]: I only want to override the `read()` method.

**Comment by TK-one** [comment]
전에도 말했지만 Service layer를 바꾸는게 아니야. Storage layer를 바꿔야지.
Service는 어떻게 조회하는지 몰라야돼. 캐싱을 하든 메모리에서 가져오든 다른 API에서 가져오든.
우리는 지금 비즈니스 로직을 건드는게 아니잔아?
board를 어떻게 조회하는지를 변경하려는거잔아. (원래 db에서만 조회하던것에서 cache를 추가)

`Cacheable~` 은 Storage layer에서 등장해야지. `CacheableBoardStorage` 라든지 `CacheableBoardRepository` 라든지
Service코드는 그대로 있는거야.

나도 최근에 작업한게 있는데 여기 코드한번 나중에 읽어봐
- https://github.com/TK-one/supply-server/tree/develop/supply-core/src/main/java/com/supply/boardcategory/adapter/out
- https://github.com/TK-one/supply-server/tree/develop/supply-core/src/main/java/com/supply/ipname/adapter/out

File: src/main/java/com/mide/gangsaeng/board/CacheableBoardReadStorage.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import io.lettuce.core.SetArgs;
+import io.lettuce.core.api.sync.RedisCommands;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@Repository
+public class CacheableBoardReadStorage implements BoardReadStorage {
+    private final BoardReadStorage db;
+    private final BoardReadStorage cache;
+    private final RedisCommands<String, String> redisCommands;
```
Comment (TK-one) [inline src/main/java/com/mide/gangsaeng/board/CacheableBoardReadStorage.java:22]: We should not use `RedisCommands` here.
This should be used only in `BoardReadCacheStorage`.

This `CacheableBoardReadStorage` focus on **how to manipulate each cache and db storage implementatons** and provide cacheable storage layer.

File: src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java [hunk 3]
```diff
@@ -48,12 +39,7 @@ public void write(BoardRequest request) {
 
     @Override
     public BoardResponse read(long id) {
-        Board board = fetchCachedBoard(id);
-
-        if (board == null) {
-            board = boardRepository.read(id);
-            cacheBoard(board);
-        }
+        Board board = boardReadStorage.read(id);
```
Comment (TK-one) [inline src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java:42]: Please handle null case here.

File: src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java [hunk 4]
```diff
@@ -20,23 +22,30 @@ public class BoardServiceImpl implements BoardService {
...
+
+        Board board = boardMapper.boardRequestToBoard(request);
+        boardRepository.write(board);
     }
 
     @Override
     public BoardResponse read(long id) {
         Board board = boardRepository.read(id);
+        Optional.ofNullable(board)
+                .orElseThrow(() -> new BoardNotFoundException(ErrorCode.BOARD_NOT_FOUND));
```
Comment (TK-one) [inline src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java:48]: Why did you use Optional here?
We can just check null directly

File: src/main/java/com/mide/gangsaeng/board/CacheableBoardRepository.java [hunk 5]
```diff
@@ -26,8 +23,8 @@ public CacheableBoardRepository(
     }
 
     @Override
-    public void write(BoardRequest request) {
-        db.write(request);
+    public void write(Board board) {
+        db.write(board);
```
Comment (TK-one) [inline src/main/java/com/mide/gangsaeng/board/CacheableBoardRepository.java:27]: What about also writing to redis here?

File: src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java (hunk 4 above)
Comment (kanghowoo) [reply src/main/java/com/mide/gangsaeng/board/BoardServiceImpl.java:48]: I thought I could express null and exception handling with concise code...

File: src/main/java/com/mide/gangsaeng/board/CacheableBoardRepository.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            BoardCacheRepositoryImpl cache) {
+        this.db = db;
+        this.cache = cache;
+    }
+
+    @Override
+    public void write(Board board) {
+        Board boardDataToBeStored = board.toBuilder()
+                                          .createdAt(LocalDateTime.now().withNano(0))
+                                          .updatedAt(LocalDateTime.now().withNano(0))
```
Comment (TK-one) [inline src/main/java/com/mide/gangsaeng/board/CacheableBoardRepository.java:31]: I think we don't need to call `withNanos` here. why did you call it?

File: src/main/resources/mybatis/mappers/board/boardMapper.xml [hunk 7]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 <!DOCTYPE mapper PUBLIC "-//mybatis.org//DTD Mapper 3.0//EN"
   "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
-<mapper namespace="com.mide.gangsaeng.board.BoardMapper">
-  <insert id="write" parameterType="com.mide.gangsaeng.board.BoardRequest">
+<mapper namespace="com.mide.gangsaeng.board.BoardMybatisMapper">
+  <insert id="write" parameterType="com.mide.gangsaeng.board.Board" useGeneratedKeys="true" keyProperty="id">
     INSERT INTO boards (title, content, user_id) values (#{title}, #{content}, #{userId});
```
Comment (TK-one) [inline src/main/resources/mybatis/mappers/board/boardMapper.xml:6]: Where is insert query for createdAt and updatedAt ?

File: src/main/resources/mybatis/mappers/board/boardMapper.xml [hunk 8]
```diff
@@ -3,7 +3,8 @@
   "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
 <mapper namespace="com.mide.gangsaeng.board.BoardMybatisMapper">
   <insert id="write" parameterType="com.mide.gangsaeng.board.Board" useGeneratedKeys="true" keyProperty="id">
-    INSERT INTO boards (title, content, user_id) values (#{title}, #{content}, #{userId});
+    INSERT INTO boards (title, content, user_id, created_at, updated_at)
+    values (#{title}, #{content}, #{userId}, #{createdAt}, #{updatedAt});
```
Comment (TK-one) [inline src/main/resources/mybatis/mappers/board/boardMapper.xml:7]: Can you print the actual query when application insert data?

Comment (kanghowoo) [reply src/main/resources/mybatis/mappers/board/boardMapper.xml:7]: image

**Review by TK-one — APPROVED** [review]
LGTM

**Pull request merged**